sklearn/inspection/permutation_importance.py

```diff
@@ -0,0 +1,82 @@
+"""Permutation importance for estimators"""
+import numpy as np
+
+from ..utils import check_random_state
+from ..compose._column_transformer import _check_X
+from ..metrics import check_scoring
+
+
+def permutation_importance(estimator, X, y, scoring=None, n_rounds=1,
+                           random_state=None):
+    """Permutation importance for feature evaluation. [BRE]_
+
+    The permutation importance of a feature is calculated as follows. First,
+    the estimator is trained on a training set. Then a baseline metric, defined
+    by ``scoring``, is evaluated on a (potentially different) dataset defined
+    by the ``X`` parameter. Next, a feature column from the validation set is
+    permuted and the metric is evaluated again. The permutation importance is
+    defined to be the difference between the baseline metric and metric from
+    permutating the feature column.
+
+    Read more in the :ref:`User Guide <permutation_importance>`.
+
+    Parameters
+    ----------
+    estimator : object
+        An estimator that has already been `fit` and is compatible with
+        ``scorer``.
+
+    X : array-like or DataFrame, shape = (n_samples, n_features)
+        Data on which permutaiton importance will be computed.
+
+    y : array-like, shape = (n_samples, ...)
+        Targets for supervised learning.
+
+    scoring : string, callable or None, optional (default=None)
+        Scorer to use. It can be a single
+        string (see :ref:`scoring_parameter`) or a callable (see
+        :ref:`scoring`). If None, the estimator's default scorer is used.
+
+    n_rounds : int, optional (default=1)
+        Number of times to permute a feature.
+
+    random_state : int, RandomState instance or None, optional, default None
+        Pseudo-random number generator to control the permutations.
+        See :term:`random_state`.
+
+    Returns
+    -------
+    importances : array, shape (n_features, n_rounds)
+        Permutation importance scores.
+
+    References
+    ----------
+    .. [BRE] Breiman, L. Machine Learning (2001) 45: 5.
+        https://doi.org/10.1023/A:1010933404324
+    """
+    X = _check_X(X)
+    random_state = check_random_state(random_state)
+    scorer = check_scoring(estimator, scoring=scoring)
+    scores = np.empty(shape=(X.shape[1], n_rounds), dtype=np.float)
+
+    if hasattr(X, 'iloc'):  # pandas dataframe
+        # Need to copy in case where the dataframe is a view
+        X = X.copy()
```

Comment:
Why only copy when it's a df? Numpy arrays can also be views.

Reply:
do dataframes actually fail when writing to a view?

Reply:
When `X` is already a view, as returned by `train_test_split`, there will be a `SettingWithCopyWarning` warning raised when this modifies the dataframe in place.

Reply:
Shouldn't we copy in all cases anyway?

Df or not, it X is a view, it might be used by some other worker at the same time. Since we're modifying it inplace this can have bad consequences